src/dotty/tools/dotc/core/Types.scala

```diff
@@ -3213,6 +3232,55 @@ object Types {
     def apply(tp: Type) = tp
   }
 
+  abstract class ApproximatingTypeMap(implicit ctx: Context) extends TypeMap { thisMap =>
```

Comment:
Interesting abstraction, do you think we could implement `TypeAssigner#avoid` using it?